<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Harmonise list of GWAS data to have same variants, same effect allele, and to remove duplicate snps — harmoniseGWAS • MGflashfm</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Harmonise list of GWAS data to have same variants, same effect allele, and to remove duplicate snps — harmoniseGWAS"><meta property="og:description" content="Harmonise list of GWAS data to have same variants, same effect allele, and to remove duplicate snps"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MGflashfm</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/MGflashfm.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Harmonise list of GWAS data to have same variants, same effect allele, and to remove duplicate snps</h1>
    
    <div class="hidden name"><code>harmoniseGWAS.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Harmonise list of GWAS data to have same variants, same effect allele, and to remove duplicate snps</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">harmoniseGWAS</span><span class="op">(</span></span>
<span>  <span class="va">obsgwas</span>,</span>
<span>  minMAF <span class="op">=</span> <span class="fl">0.005</span>,</span>
<span>  minINFO <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>  beta_colname <span class="op">=</span> <span class="st">"beta"</span>,</span>
<span>  se_colname <span class="op">=</span> <span class="st">"SE"</span>,</span>
<span>  snpID_colname <span class="op">=</span> <span class="st">"rsID"</span>,</span>
<span>  EA_colname <span class="op">=</span> <span class="st">"EA"</span>,</span>
<span>  NEA_colname <span class="op">=</span> <span class="st">"NEA"</span>,</span>
<span>  EAfreq_colname <span class="op">=</span> <span class="st">"EAF"</span>,</span>
<span>  BP_colname <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>  pvalue_colname <span class="op">=</span> <span class="st">"p_value"</span>,</span>
<span>  INFO_colname <span class="op">=</span> <span class="st">"INFO"</span>,</span>
<span>  N_colname <span class="op">=</span> <span class="st">"N"</span>,</span>
<span>  minNprop <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>obsgwas</dt>
<dd><p>list of GWAS data.frames with column names as specified in the other arguments</p></dd>


<dt>minMAF</dt>
<dd><p>only variants with MAF &gt; minMAF in all GWAS are retained; default is 0.005</p></dd>


<dt>minINFO</dt>
<dd><p>only variants with INFO &gt; 0.4 in all GWAS are retained; default is 0.4</p></dd>


<dt>beta_colname</dt>
<dd><p>text column name for the effect estimates in each obsGWAS data.frame; default "BETA"</p></dd>


<dt>se_colname</dt>
<dd><p>text column name for the effect estimate standard errors in each obsGWAS data.frame; default "SE"</p></dd>


<dt>snpID_colname</dt>
<dd><p>text column name for the effect estimates in each obsGWAS data.frame; default "SNP"</p></dd>


<dt>EA_colname</dt>
<dd><p>text column name for the effect allele in each obsGWAS data.frame; default "EA"</p></dd>


<dt>NEA_colname</dt>
<dd><p>text column name for the non-effect allele in each obsGWAS data.frame; default "NEA"</p></dd>


<dt>EAfreq_colname</dt>
<dd><p>text column name for the effect allele frequency (EAF) in each obsGWAS data.frame; default "EAF"</p></dd>


<dt>BP_colname</dt>
<dd><p>text column name for the base-pair position in each obsGWAS data.frame; default "BP"</p></dd>


<dt>pvalue_colname</dt>
<dd><p>text column name for the p-value in each obsGWAS data.frame; default "p_value"</p></dd>


<dt>INFO_colname</dt>
<dd><p>text column name for the INFO score in each obsGWAS data.frame; default "INFO"</p></dd>


<dt>N_colname</dt>
<dd><p>text column name for the number of individuals with the variant measured in each obsGWAS data frame; default "N".  If there is no sample size column, then there will be no filtering using Nprop</p></dd>


<dt>minNprop</dt>
<dd><p>Use this for meta-analysis summary statistics - for each trait only variants with N/max(N) &gt; minNprop are retained; default 0.80 (e.g. each variant must be measured in at least 80% of the meta-analysis individuals)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>outputs the list of input GWAS data such that they contain the same variants, have the same effect allele (flip where needed),
and duplicates (by base-pair position) are removed, retaining the variant with highest INFO score</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Jenn Asimit</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jenn Asimit, Feng Zhou.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

