<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ MGflashfm</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MGflashfm</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/MGflashfm.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="FLASHFMwithJAMd.html">FLASHFMwithJAMd()</a></code> </p>
        </td>
        <td><p>Wrapper for flashfm Multi-Trait Fine-Mapping with JAM  - this is the dynamic number of max causal variant version</p></td>
      </tr><tr><td>
          <p><code><a href="JAMdynamic.html">JAMdynamic()</a></code> </p>
        </td>
        <td><p>Expanded version of JAM (a single-trait fine-mapping approach) that first runs on thinned SNPs and then expands models on tag SNPs
This version starts at a low upper bound for max causal variants and decides on max upper bound based on data</p></td>
      </tr><tr><td>
          <p><code><a href="LDqc.html">LDqc()</a></code> </p>
        </td>
        <td><p>Quality control of dosage genotype matrix prior to LD calculation</p></td>
      </tr><tr><td>
          <p><code><a href="MGFLASHFMwithJAM.html">MGFLASHFMwithJAM()</a></code> </p>
        </td>
        <td><p>Wrapper for Multi-group multi-trait fine-mapping, using FLASHFMwithJAM</p></td>
      </tr><tr><td>
          <p><code><a href="MGFMwithJAM.html">MGFMwithJAM()</a></code> </p>
        </td>
        <td><p>Wrapper for credible sets from Multi-Group Single-Trait Fine-Mapping with JAM</p></td>
      </tr><tr><td>
          <p><code><a href="MGflashfm.html">MGflashfm()</a></code> </p>
        </td>
        <td><p>Multi-group  multi-trait fine-mapping, using output from FLASHFMwithJAM or FLASHFMwithFINEMAP; gives both multi- and single-trait results from flashfm object</p></td>
      </tr><tr><td>
          <p><code><a href="MGflashfmRET.html">MGflashfmRET()</a></code> </p>
        </td>
        <td><p>Multi-group  multi-trait fine-mapping, using output from FLASHFMwithJAM or FLASHFMwithFINEMAP; gives both multi- and single-trait results from flashfm object</p></td>
      </tr><tr><td>
          <p><code><a href="MGfm.html">MGfm()</a></code> </p>
        </td>
        <td><p>Multi-Group Single-Trait Fine-Mapping credible sets</p></td>
      </tr><tr><td>
          <p><code><a href="Nall.html">Nall</a></code> </p>
        </td>
        <td><p>List of samples size vectors for two simulated traits in two groups</p></td>
      </tr><tr><td>
          <p><code><a href="Neff.html">Neff()</a></code> </p>
        </td>
        <td><p>Approximate effective sample size from GWAS of related samples (copied from flashfm)</p></td>
      </tr><tr><td>
          <p><code><a href="alignGWAS.html">alignGWAS()</a></code> </p>
        </td>
        <td><p>Flip GWAS variants to align with SNP correlation matrix (LD) from reference panel or in-sample</p></td>
      </tr><tr><td>
          <p><code><a href="allcredsetsPP.html">allcredsetsPP()</a></code> </p>
        </td>
        <td><p>Construct a credible set for each trait and under each of single and multi-trait fine-mapping, and provide SNP PP details</p></td>
      </tr><tr><td>
          <p><code><a href="corX.list.html">corX.list</a></code> </p>
        </td>
        <td><p>List of SNP correlation matrices for two groups</p></td>
      </tr><tr><td>
          <p><code><a href="covY.list.html">covY.list</a></code> </p>
        </td>
        <td><p>Covariance matrices for two simulated traits in two groups</p></td>
      </tr><tr><td>
          <p><code><a href="cvs.html">cvs</a></code> </p>
        </td>
        <td><p>Vector of rsids for causal variants of two simulated traits in two groups</p></td>
      </tr><tr><td>
          <p><code><a href="g2t1CS.html">g2t1CS</a></code> </p>
        </td>
        <td><p>Output from running JAMdynamic on Group 2, Trait 1 simulated dataset in this package</p></td>
      </tr><tr><td>
          <p><code><a href="groupmultiU.html">groupmultiU()</a></code> </p>
        </td>
        <td><p>Group SNPs; adapted from group.multi by Chris Wallace</p></td>
      </tr><tr><td>
          <p><code><a href="gwas.list.html">gwas.list</a></code> </p>
        </td>
        <td><p>List of GWAS results for two traits in two groups</p></td>
      </tr><tr><td>
          <p><code><a href="harmoniseGWAS.html">harmoniseGWAS()</a></code> </p>
        </td>
        <td><p>Harmonise list of GWAS data to have same variants, same effect allele, and to remove duplicate snps</p></td>
      </tr><tr><td>
          <p><code><a href="makeSNPgroups2U.html">makeSNPgroups2U()</a></code> </p>
        </td>
        <td><p>Make two sets of SNP groups using fine-mapping information from all of the traits using two sets of results and maps the names between them</p></td>
      </tr><tr><td>
          <p><code><a href="makeSNPgroupsU.html">makeSNPgroupsU()</a></code> </p>
        </td>
        <td><p>Make SNP groups using fine-mapping information from all of the traits</p></td>
      </tr><tr><td>
          <p><code><a href="mgmtCS.html">mgmtCS</a></code> </p>
        </td>
        <td><p>Output from running MGFLASHFMwithJAM on the simulated dataset in this package</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jenn Asimit, Feng Zhou.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

